Support ref property inside textInputProps

jakub-trzebiatowski opened the pull request:
Ensure that it doesn't interfere with the internal reference to the text input.

This changes are required for the sake of https://github.com/Expensify/App/pull/16433 (currently a draft), which aims to close https://github.com/Expensify/App/issues/14783.

jakub-trzebiatowski commented on src/index.js:
```diff
@@ -12,6 +12,12 @@ const IOS_MODAL_HEIGHT = 262;
     static propTypes = {
+        // A platform-dependent reference to an element that exposes a best-effort `focus` method that doesn't involve
+        // opening the picker
+        focusRef: PropTypes.oneOfType([
```
I picked a custom name for the ref, as forwarding the real ref broke literally [the whole testsuite](https://github.com/Expensify/react-native-picker-select/blob/84ee97dec11c2e65609511eb5a757d61bbeeab79/test/test.js#L98). It's all built on the assumption that the `ref` actually points to the actual `RNPickerSelect` component.

jakub-trzebiatowski replied:
Outdated

parasharrajat commented on src/index.js:
```diff
@@ -12,6 +12,12 @@ const IOS_MODAL_HEIGHT = 262;
 
 export default class RNPickerSelect extends PureComponent {
     static propTypes = {
+        // A platform-dependent reference to an element that exposes a best-effort `focus` method that doesn't involve
+        // opening the picker
+        focusRef: PropTypes.oneOfType([
+            PropTypes.func,
+            PropTypes.shape({ current: PropTypes.instanceOf(React.Component) }),
+        ]),
```
```suggestion
```

Not needed.

jakub-trzebiatowski replied:
Of course I understand your suggestion to drop an extra ref property (`focusRef`; but name is not crucial) and rely on `pickerProps.ref` / `textInputProps.ref`.

But it has consequences, right? In the bigger picture, you actually mean that it would be better to do our own platform switching on top of `react-native-picker-select/picker`. Do I understand that correctly?

```js
if (Platform.OS == 'web') {
  return <RNPicker pickerProps={{ref: /* ... */}} /* ... */ />;
} else {
  return <RNPicker textInputProps={{ref: /* ... */}} /* ... */ />;
}
```

...or something among that lines.

Unless I'm missing something.

I decided to put changes here, as I'm pretty sure that that's what author of the approach had in mind. But the extra switch on the app-side is also totally fine for me.

jakub-trzebiatowski replied:
@parasharrajat [Relevant thread](https://github.com/Expensify/App/pull/16433/files#r1146432817) on the use-site

parasharrajat replied:
I don't think there are any consequences here. Allowing to pass ref's through two ways is confusing.  I think it is hard to argue the philosophy of the lib but we can just do minimal changes that are required for us or a blocker for us. You could be right to use `focusref` logic and it could be serving a better purpose.

the main blocker for us is the overwriting ref for TextInput. we don't have to extend this to the focus concept that is only relevant in our App.

jakub-trzebiatowski replied:
Okey, I can see we're in agreement; I will adjust both sides of this solution

parasharrajat commented on src/index.js:
```diff
@@ -67,6 +73,7 @@ export default class RNPickerSelect extends PureComponent {
     };
 
     static defaultProps = {
+        focusRef: undefined,
```
```suggestion
```

parasharrajat commented on src/index.js:
```diff
@@ -618,3 +630,11 @@ export default class RNPickerSelect extends PureComponent {
 }
 
 export { defaultStyles };
+
+function setRef(ref, el) {
+    if (typeof ref === 'function') {
+        ref(el);
+    } else if (ref) {
+        ref.current = el;
+    }
+}
```
```suggestion
```

jakub-trzebiatowski replied:
I don't really understand why inlining this function improves readability. This is basic DRY.

```
      if(typeof textInputProps.ref === 'function'){
       textInputProps.ref(el);
      }else{
      textInputProps.ref.current = el;
      }
```

In practice, such code is more prone to mistakes and requires extra brain power to ensure that all three `textInputProps.ref` are indeed the same expression.

parasharrajat replied:
> This is basic DRY

It is unnecessary to put it inside a function if it is only used once.  Also, it does not seem to serve a big difference. Inline code is more readable.

jakub-trzebiatowski replied:
I agree neither with the first (i.e that functions are only for things used more than once) nor with the second

But I assume that your word here is final; I'll inline

parasharrajat replied:
No, my word is not final. This is a no-blocker for me. But we can at least put the function at the top. ESLint will throw `Function used before is defined`...

jakub-trzebiatowski replied:
I already followed your suggestion, so there's no need to jump back and forth. But I appreciate that you were open to my perspective.

parasharrajat commented on src/index.js:
```diff
@@ -590,6 +601,7 @@ export default class RNPickerSelect extends PureComponent {
                     onValueChange={this.onValueChange}
                     selectedValue={selectedItem.value}
                     {...pickerProps}
+                    ref={this.props.focusRef}
```
```suggestion
```

I didn't understand this duplicate ref assignment.

jakub-trzebiatowski replied:
It's not duplicate, it's another approach where `RNPickerSelect` is controlling the refs and they aren't overridable. But I already understand that you disagree with this approach.

It all comes down to [this thread](https://github.com/Expensify/react-native-picker-select/pull/6#pullrequestreview-1354740078)

parasharrajat replied:
I am thinking that we should do minimal changes first.

parasharrajat commented on src/index.js:
```diff
@@ -563,6 +573,7 @@ export default class RNPickerSelect extends PureComponent {
                     onValueChange={this.onValueChange}
                     selectedValue={selectedItem.value}
                     {...pickerProps}
+                    ref={this.props.focusRef}
```
```suggestion
```

parasharrajat commented on src/index.js:
```diff
@@ -436,9 +443,12 @@ export default class RNPickerSelect extends PureComponent {
+                    ref={(el) => {
+                        this.setInputRef(el);
+                        setRef(this.props.focusRef, el);
+                    }}
```
```suggestion
                    ref={(el) => {
                        this.setInputRef(el);
                        
                        if(!textInputProps.ref){
                           return;
                        } 
                        
                        if(typeof textInputProps.ref === 'function'){
                         textInputProps.ref(el);
                        }else{
                        textInputProps.ref.current = el;
                        }
                    }}
```

parasharrajat reviewed (changes requested):
we have to only fix text input ref forwarding as it overwrites the internal ref prop.

parasharrajat commented:
Please add a description to the PR. Explain Why are we doing these changes?

parasharrajat commented:
Please do not force push. we don't prefer force pushes. Also, commits needs to be GPG signed.

jakub-trzebiatowski commented:
> Please do not force push. we don't prefer force pushes.

That means that every typo, discarded idea or non-working approach lands in the repository? 😶 That's unusual.

I forced pushed to undo the force push.

I'll figure out the GPG signing.

jakub-trzebiatowski committed:
Support ref property inside textInputProps

...as a replacement for the approach with exposing a custom reference
property. Ensure that it doesn't interfere with the internal reference
to the text input.

Inline setRef to (arguably) increase readability.

jakub-trzebiatowski committed:
Expose the focusRef reference

jakub-trzebiatowski commented:
@parasharrajat One last force push, because otherwise I couldn't add GPG signing

parasharrajat reviewed (approved):
LGTM. I will ask someone to merge it when I have tested the original PR.

parasharrajat commented:
@cubuspl42 We can close this PR.
